### PR TITLE
.FEC F3X Summary Row

### PR DESCRIPTION
**Pull request by toddlees**
## Summary

- Introduces Celery task to serialize the F3X summary into a csv row.  As we implement more parts of the .FEC file, we will compile these rows together.  To fire off the task, POST to `api/v1/web-services/create-dot-fec/` with a body like this: `{"report_id": 26}`.  
image

- The celery worker logs will show what's happening on the serialization side:
image
The last line shows the csv row that was created. This example doesn't have a lot of fields defined.

### CODE REVIEW

File: django-backend/fecfiler/web_services/tasks.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+F3X_SCHEMA = validate.get_schema("F3X")
+F3X_PROPERTIES = F3X_SCHEMA.get("properties", {}).items()
+F3X_COLUMNS = {v.get("fec_spec", {}).get("COL_SEQ", None): k for k, v in F3X_PROPERTIES}
+F3X_NUMBER_OF_COLUMNS = max(F3X_COLUMNS.keys())
+SERIALIZERS = {
+    models.BooleanField: lambda b, o, fn: "X" if b else "",
+    models.DateField: lambda d, o, fn: d.strftime("%Y%m%d") if d else "",
+    models.ForeignKey: lambda fo, o, fn: getattr(o, fn + "_id", ""),
+    None: lambda v, o, fn: str(v) if v else "",
+}
```
Comment (Elaine-Krauss-TCG) [inline django-backend/fecfiler/web_services/tasks.py:19]: Is there anything that can be done to make this section a little more readable?

File: django-backend/fecfiler/web_services/tasks.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+def serialize_f3x_summary(report_id):
+    f3x_summary_result = F3XSummary.objects.filter(id=report_id)
+    if f3x_summary_result.exists():
+        logger.info(f"serializing f3x summary: {report_id}")
+        f3x_summary = f3x_summary_result.first()
+        row = [
+            serialize_field(f3x_summary, F3X_COLUMNS[column_index], F3XSummary)
+            for column_index in range(1, F3X_NUMBER_OF_COLUMNS + 1)
+        ]
+        return ",".join(row)
```
Comment (Elaine-Krauss-TCG) [inline django-backend/fecfiler/web_services/tasks.py:37]: *chef's kiss*

**Review by Elaine-Krauss-TCG — APPROVED** [review]
Looks really good :+1:

File: django-backend/fecfiler/web_services/tasks.py (hunk 1 above)
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/web_services/tasks.py:19]: E.G. `models.ForeignKey: lambda fo, o, fn: getattr(o, fn + "_id", "")`

I _think_ that "fo" is the foreign key, and "o" is the resulting object, but I would not be able to tell you why it's trying to get `fn + "_id"` from "o"

**Review by dheitzer — APPROVED** [review]

File: django-backend/fecfiler/web_services/tasks.py (hunk 1 above)
Comment (toddlees) [reply django-backend/fecfiler/web_services/tasks.py:19]: great suggestion!  I left my code golf hat on while writing that.  updating now

File: django-backend/fecfiler/web_services/dot_fec_serializer.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+"""A map of model field types to their serializers.
+Pass the model instance and field name into the serializer to
+get a string representation in the FEC standard
+"""
+FIELD_SERIALIZERS = {
+    models.BooleanField: boolean_serializer,
+    models.DateField: date_serializer,
+    models.ForeignKey: foreign_key_serializer,
+    None: default_serializer,
```
Comment (Elaine-Krauss-TCG) [inline django-backend/fecfiler/web_services/dot_fec_serializer.py:45]: RIP lambdas.  They were a very elegant solution, but this is really cool in its own right and should be much, much easier to expand on.

**Review by Elaine-Krauss-TCG — APPROVED** [review]
:100:

File: django-backend/fecfiler/web_services/dot_fec_serializer.py (hunk 3 above)
Comment (toddlees) [reply django-backend/fecfiler/web_services/dot_fec_serializer.py:45]: yeah i think you were right to point out that they were getting unreadable.

**Review by mjtravers — CHANGES_REQUESTED** [review]
Looks good except for the 2 code smells. Would you make a const for the "api/v1" string and for the other smell we can create a simple child Exception class to be used across all site apps

**Review by mjtravers — APPROVED** [review]

**Pull request merged**